Сделал ДЗ №3

4nt0n64r opened the pull request:
- Сделать в активити фрагмент с RecyclerView.
- Отобразить в RecyclerView список наименований заметок
- Создать второй фрагмент для отображения информации о заметке (наименование заметки, текст заметки и, например, дата заметки)
- Реализовать показ деталей заметки на втором фрагменте при нажатии на пункт RecyclerView

4nt0n64r committed:
Сделал ДЗ №3

- Сделать в активити фрагмент с RecyclerView.
- Отобразить в RecyclerView список наименований заметок
- Создать второй фрагмент для отображения информации о заметке (наименование заметки, текст заметки и, например, дата заметки)
- Реализовать показ деталей заметки на втором фрагменте при нажатии на пункт RecyclerView

BSCAndroid commented:
Пока что не по коду, а по работе приложения: 
1. При клике на ячейку нужно запускать экран с редактированием заметки (должен присутствовать шаринг и сохранение, сохранение можно замокать появлением тоста или записывать в лог)

4nt0n64r committed:
Исправления замечаний

При клике на ячейку нужно запускать экран с редактированием заметки (должен присутствовать шаринг и сохранение, сохранение можно замокать появлением тоста или записывать в лог)

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/MVPInterfaces.kt:
```diff
@@ -10,27 +14,118 @@ interface ViewInterface {
     /**
      *  Обработка клика на кнопку "Save"
      *
      *  @param message - сообщение
```
добавление в kd символа "-" добавляет сайд эффект при генерации UI доки в IDE: добавляется элемент, именуемый bullet. Посмотри это в IDE, чтоб убедиться. Во всех файлах лучше это удалить.

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/MainActivity.kt:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+
+    private fun Activity.hideKeyboard() {
```
подобные вещи лучше вынести в отдельный файл с расширениями, чтоб был глобальный доступ
в случае со скрытием клавиатуры - аналогично

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/list/ListFragment.kt:
```diff
@@ -0,0 +1,67 @@
+ */
+class ListFragment : Fragment(), ListView {
+
+    lateinit var binding: FragmentListBinding
```
в случае с фрагментами биндинг нужно делать nullable: 
https://developer.android.com/topic/libraries/view-binding

и в случае нашего проекта: отказаться от использования !! в коде

PS: комментарий работает для всех фрагментов

4nt0n64r replied:
Если отказаться получать binding через !! из _binding, то наступает nullable ад. Даже в доках от гугла так получают же.  Да и куча этих ? не добавляют читаемости. Если биндинг будет null каким-то образом, то лучше уж падать чем так)

BSCAndroid replied:
что значит nullable ад? И что за куча "?"?
большинство проектов из прода не используют "!!". 
ты можешь всегда использовать скоуп функции для nullable объектов, тогда и нескольких в ряд знаков вопросов не будет.

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/list/ListFragment.kt:
```diff
@@ -0,0 +1,67 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        presenter = ListPresenterImpl(this)
```
презентер должен быть синхронизирован по ЖЦ фрагмента в отношении его View. То есть когда мы уничтожаем вью фрагмента необходимо занулить презентер

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/list/ListFragment.kt:
```diff
@@ -0,0 +1,67 @@
+
+    private fun setupRecyclerView() {
+        adapter = NotesAdapter { note -> noteClick(note) }
+        binding.noteList.layoutManager = LinearLayoutManager(context)
```
можно использовать скоуп функцию apply:
binding.noteList.apply { ... }

подробности тут: 
https://kotlinlang.ru/docs/reference/scope-functions.html

4nt0n64r replied:
Я вообще сначала так и хотел через apply сделать, но если применить apply - RecyclerView не заполняется.

BSCAndroid replied:
использовал "this@ListFragment"?

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/note/NoteFragment.kt:
```diff
@@ -0,0 +1,91 @@
+ */
+class NoteFragment : Fragment(), NoteView {
+
+    lateinit var binding: FragmentNoteBinding
```
nullable FragmentNoteBinding

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/note/NoteFragment.kt:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    override fun showNote(note: NoteUI) {
+        binding.noteHeader.setText(note.header)
```
когда binding будет nullable, то можно через ?.let { ... } переписать эти 3 строчки

BSCAndroid commented:
Уточнение требований: 
1. AboutActivity вызывается только экрана со списком заметок

4nt0n64r committed:
Исправления замечаний по коду

4nt0n64r committed:
Сделал биндинг нуллабельным

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/list/ListFragment.kt:
```diff
@@ -0,0 +1,89 @@
+        savedInstanceState: Bundle?
+    ): View? {
+        binding = FragmentListBinding.inflate(layoutInflater)
+        return binding.let { it?.root }
```
излишние и запутывающие конструкции, можно сделать так: 

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = FragmentNoteBinding.inflate(layoutInflater).also {
        binding = it
    }.root

4nt0n64r replied:
Мне не дает так сделать `Type mismatch: inferred type is FragmentNoteBinding but FragmentListBinding? was expected`

4nt0n64r replied:
Немного затупил, теперь поправил

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/list/ListFragment.kt:
```diff
@@ -0,0 +1,89 @@
+
+    override fun onDestroy() {
+        super.onDestroy()
+        binding = null
```
binding зануляться должен не тут, а в `onDestroyView`

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/list/ListFragment.kt:
```diff
@@ -0,0 +1,89 @@
+        setupRecyclerView()
+        presenter?.getNotes()
+
+        binding.let {
```
почему сделал так (в этом коммите), а не через 

```
binding?.let {
// тут можно обращаться к binding без ? через it
}
```

или 

```
binding?.let { binding ->
// тут получается именованная лямбда
}
```
?

4nt0n64r replied:
Не использовал let в таком ключе, буду теперь знать

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/note/NoteFragment.kt:
```diff
@@ -0,0 +1,105 @@
+        savedInstanceState: Bundle?
+    ): View? {
+        binding = FragmentNoteBinding.inflate(layoutInflater)
+        return binding.let { it?.root }
```
тут такой же момент, описанный выше

4nt0n64r replied:
Мне не дает так сделать `Type mismatch: inferred type is FragmentNoteBinding but FragmentListBinding? was expected`

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/note/NoteFragment.kt:
```diff
@@ -0,0 +1,105 @@
+            presenter?.setNote(this.requireArguments().getSerializable(NOTE_ITEM_BUNDLE) as NoteUI)
+        }
+
+        binding.let {
```
тут такой же момент, описанный выше (и ниже по всему файлу)

BSCAndroid commented on app/src/main/java/com/a4nt0n64r/notes/main/fragments/note/NoteFragment.kt:
```diff
@@ -0,0 +1,105 @@
+
+    override fun onDestroy() {
+        super.onDestroy()
+        binding = null
```
тут такой же момент, описанный выше

4nt0n64r committed:
Фикс замечаний по применению let
